### ISSUE TITLE
Way to Systematically Fix Issues with "Series" Column

### ISSUE THREAD

**Issue by jacole3** [mrcaseb/nflfastR#130]
EDIT: I've added more code to systematically fix the series_result variable as well, since it originally has a lot of NAs. It can be seen at the same link below in this post. I'll spare the screenshots here since this post is already long, but happy to discuss with anyone in the comments or privately.

Note: The following data is designed to work with NFLFastR version 3.1.1. Descriptions have been added with the "#" sign to help increase reproducibility. It is possible the specific play_id numbers will no longer be accurate if/when a new NFLFastR version is released.

If you've done any work that involves the "series" column in NFLFastR, you might have noticed some ordering issues with it. It's not quite the shitshow that the "drive" and "fixed_drive" column are (see issues #36 and #67 ), but there are some mistakes. Many of the mistakes come from NFLFastR erroneously counting PATs as their own series, but there are also some miscellaneous issues, like these few examples below: 

View(pbp %>% filter(game_id == "2000_17_CIN_PHI", qtr == 4))

<img width="850" alt="Series Wrong Example 1 CIN PHI" src="https://user-images.githubusercontent.com/39979582/96971835-8909c400-14ca-11eb-9667-5b5d077fa26a.png">

View(pbp %>% filter(game_id == "2004_09_PHI_PIT", qtr == 2))

<img width="850" alt="Series Wrong Example 2 PHI PIT" src="https://user-images.githubusercontent.com/39979582/96971848-8dce7800-14ca-11eb-9655-44d8e09743fb.png">

View(pbp %>% filter(game_id == "2004_15_WAS_SF", qtr == 1))

<img width="845" alt="Series Wrong Example 3 WAS SF" src="https://user-images.githubusercontent.com/39979582/96971855-932bc280-14ca-11eb-8104-26675db96d3e.png">

As was the case with the "drive" column, we simply can't discount these plays as mistakes coming from the late 90s/early 2000s era, when NFL data tracking wasn't as accurate. Issues show up throughout the past 20 years.

Going through every single game by hand and manually correcting the series numbers is an unreasonable task, but fortunately, I've put together a new system that systematically corrects the series numbers by creating a new variable. The full code can be seen here: https://github.com/jacole3/Fixing-NFLFastR-Series-Column

I've gone through more than 10 games, and I've not seen any errors with the "NewSeries" column. For frame of reference, here are the three examples we looked at earlier:

<img width="886" alt="Fixed NewSeries Example 1 CIN PHI" src="https://user-images.githubusercontent.com/39979582/96972074-e1d95c80-14ca-11eb-996a-214da25ebc4d.png">

<img width="885" alt="Fixed NewSeries Example 2 PHI PIT" src="https://user-images.githubusercontent.com/39979582/96972086-e56ce380-14ca-11eb-920f-d555cb3969df.png">

<img width="887" alt="Fixed NewSeries Example 3 WAS SF" src="https://user-images.githubusercontent.com/39979582/96972098-e9990100-14ca-11eb-9ecb-bdf0f329ae77.png">

I hope this is helpful to all other members of the NFLFastR community who are tired of seeing data entry errors in the repository. If anyone does spot mistakes with the "NewSeries" column, feel free to either reply here, or contact me directly at @ColeJacobson32 on Twitter or jacole@sas.upenn.edu.

**Comment by saiemgilani** [comment]
Not my repo or my issue, I was just taking a look at the series data code:
[https://github.com/mrcaseb/nflfastR/blob/master/R/helper_add_series_data.R](helper_add_series_data.R)
```      
new_series = dplyr::if_else(
# a new drive
.data$fixed_drive != dplyr::lag(.data$fixed_drive) |
# or a first down on the prior play
dplyr::lag(.data$first_down_rush == 1) 
#^^^^^^^^^^^^^
```
**^^^this piece right here is probably not right**
@mrcaseb 

```
 | dplyr::lag(.data$first_down_pass) == 1 | 
dplyr::lag(.data$first_down_penalty) == 1 |
# or the first play
.data$row == 1,
1, 0
),

```

**Comment by jacole3** [comment]
Interesting insight @saiemgilani , definitely suggests that fixed_drive is the root of the issues. Which I took a look into in Issue #67 (https://github.com/mrcaseb/nflfastR/issues/67). So hopefully my work both in that thread and this one can be helpful for people to work around the NFLFastR code bugs.

**Comment by saiemgilani** [comment]
Well that's not related to your issue with fixed drive, that could probably be resolved with a `tidyr::fill(.data$fixed_drive, direction = "downup")`, or something similar, while NA_character_ is assigned to fixed drive for point afters. I think while that's among the potential confounders, there's also the consideration that those dplyr::lags are vectorized and probably need to be further conditioned to step over non-plays like timeouts and end periods potentially depending upon when in the process those are added. I'm store the lags/leads as columns first to be safe and then performing the comparisons for [https://github.com/saiemgilani/cfbscrapR](cfbscrapR)

**Comment by jacole3** [comment]
Right, perhaps I misspoke there. I meant that fixed_drive itself was something I looked into in #67 , not that I broke down the specific dplyr issue that you mentioned. Using my "new variables" is a sufficient workaround, but I think your thought process is really good to attack the issues at their root. Will be interesting to see what Sebastian/Ben have to say about it, and if their code gets adjusted at all as a result.

**Comment by mrcaseb** [comment]
> 
> 
> Not my repo or my issue, I was just taking a look at the series data code:
> [https://github.com/mrcaseb/nflfastR/blob/master/R/helper_add_series_data.R](helper_add_series_data.R)
> 
> ```
> new_series = dplyr::if_else(
> # a new drive
> .data$fixed_drive != dplyr::lag(.data$fixed_drive) |
> # or a first down on the prior play
> dplyr::lag(.data$first_down_rush == 1) 
> #^^^^^^^^^^^^^
> ```
> 
> **^^^this piece right here is probably not right**
> @mrcaseb
> 
> ```
>  | dplyr::lag(.data$first_down_pass) == 1 | 
> dplyr::lag(.data$first_down_penalty) == 1 |
> # or the first play
> .data$row == 1,
> 1, 0
> ),
> ```

All right I looked into this and although this is a wrong parenthesis, it makes no difference, because the logical comparison is done inside dplyr::lag() and not outside. I think we will fix it anyway because it's more intuitive to do it like @saiemgilani suggested it.

However, this might be the right time to look at series and fixed_drive (again). The biggest problem should be series increasing on PATs. I think this is an easy fix as it is happening because `first_down_rush` and `first_down_pass` are 1 on td plays.
![grafik](https://user-images.githubusercontent.com/38586519/97280001-f6f01b80-183b-11eb-9766-08a53bd0e897.png)

So this line
https://github.com/mrcaseb/nflfastR/blob/9ae4bb1951a5b4302bc0e3e83261f5bb4406af32/R/helper_add_series_data.R#L26
should be `FALSE` if `lag(touchdown) == 1`, e.g.
```r
new_series = dplyr::if_else(
        # a new drive
        .data$fixed_drive != dplyr::lag(.data$fixed_drive) |
          # or a first down on the prior play
          (dplyr::lag(.data$first_down) == 1 & dplyr::lag(.data$touchdown) == 0)) |
          # or the first play
          .data$row == 1,
        1, 0
      )
```

I used `first_down` here because we make this helper in `clean_pnp()` https://github.com/mrcaseb/nflfastR/blob/9ae4bb1951a5b4302bc0e3e83261f5bb4406af32/R/helper_additional_functions.R#L149

Any objections @guga31bb ?

**Comment by saiemgilani** [comment]
If you want to gimme a few days, I'll sort out both this and your fixed drives column. I just worked it out for cfb version. You need a two to three play look back. All the adjustments you have to make for your important metric variables, you gotta do here, whether it be accounting for timeout-end period-timeout sequences or what have you. It is inherently relying on a match to the last active play event. Your method of using lags makes me uneasy.

First defining them inside of a group_by without either an arrange after grouping or an order inside of the lags is potentially problematic. I think default sort behavior for lag or group_by may be to default to sorting by the last column in the data frame. You've also defined them within a vectorized if_else, which may not be a problem as you have written everything only one level deep, but if you were to make it any more complicated within that statement you will run into issues.

To sum up:
- we don't trust these lags
- keep your pimp sort strong and get your plays arranged and make sure they stay in that same order.
- next time you use a lag, remember _"these rows ain't loyal"_
- keep your flag columns and fills separate by duplicating the flag and filling the duplicate so you can keep an audit column.  
- profit

At least that's where I think I landed on it for cfb

**Comment by guga31bb** [comment]
I think this change to series makes sense.

For fixed_drive, I want to be **extremely** cautious before making any changes because it took a lot of time to be able to get it to replicate other sites' numbers for e.g. red zone touchdown %.

**Comment by jacole3** [comment]
Nice, glad to hear you're looking into it. For what it's worth, to clear up the confusion, fixed_drive is not my variable; fixed_drive is included in the most recent version of NFLFastR. The one I created, which is the one that heavily uses the lag() function, is called NewDrive (as seen in #67 ). With that one, I went through at least 15 games and didn't see any labeling errors, so that sample size suggests that the lag() is safe in this context.

**Comment by mrcaseb** [comment]
I hope #133 will fix series increasing on PATs. We might still have issues but this is a first step we should take to remove a lot of false series numbers
